### PR TITLE
CLOUD-2331 - EAP CD Images

### PR DESCRIPTION
**Pull request by luck3y**
https://issues.jboss.org/browse/CLOUD-2331

### CODE REVIEW

**Comment by luck3y** [comment]
@bstansberry Review please.

Note, as noted before, I'll squash this when merged.

File: secrets/sso-app-secret.json [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{
```
Comment (luck3y) [inline secrets/sso-app-secret.json]: This file should be removed and fetched directly from SSO.

File: templates/eap-cd-image-stream.json [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+                    "version": "12.0"
+                }
+            },
+            "labels": {
+                "xpaas": "1.4.10"
+            },
+            "spec": {
+                "tags": [
+                    {
+                        "name": "latest",
```
Comment (luck3y) [inline templates/eap-cd-image-stream.json]: @rcernich WDYT about this use of latest? We'd have "latest" and "12" as tags in here, with "latest" always pointing to the most recent.

File: image.yaml [hunk 3]
```diff
@@ -99,7 +99,11 @@ modules:
       repositories:
           - name: cct_module
             git:
-                  url: https://github.com/jboss-openshift/cct_module.git
+                  url: https://github.com/jboss-openshift/cct_module
```
Comment (bstansberry) [inline image.yaml]: Should the .git have been dropped?

Comment (luck3y) [reply image.yaml]: Thanks. Probably not, but concreate still seems to work. I'll fix.

**Comment by luck3y** [comment]
Note the scripts in misc have been removed, I'll add to a personal repo for testing.

File: templates/eap-cd-amq-persistent-s2i.json [hunk 4]
```diff
@@ -0,0 +1,931 @@
...
+            "value": "eap-app",
+            "required": true
+        },
+        {
+            "displayName": "Custom http Route Hostname",
+            "description": "Custom hostname for http service route.  Leave blank for default hostname, e.g.: <application-name>-<project>.<default-domain-suffix>",
+            "name": "HOSTNAME_HTTP",
+            "value": "",
+            "required": false
+        },
```
Comment (tremes) [inline templates/eap-cd-amq-persistent-s2i.json]: I think this could be removed now.

File: templates/eap-cd-amq-persistent-s2i.json [hunk 5]
```diff
@@ -0,0 +1,931 @@
...
+                        "type": "ImageChange",
+                        "imageChangeParams": {
+                            "automatic": true,
+                            "containerNames": [
+                                "${APPLICATION_NAME}-amq"
+                            ],
+                            "from": {
+                                "kind": "ImageStreamTag",
+                                "namespace": "${IMAGE_STREAM_NAMESPACE}",
+                                "name": "jboss-amq-62:1.7"
```
Comment (tremes) [inline templates/eap-cd-amq-persistent-s2i.json]: I think this is too old. We should go with AMQ 6.3

File: templates/eap-cd-amq-persistent-s2i.json [hunk 6]
```diff
@@ -0,0 +1,931 @@
...
+                            "deploymentConfig": "${APPLICATION_NAME}-amq",
+                            "application": "${APPLICATION_NAME}"
+                        }
+                    },
+                    "spec": {
+                        "terminationGracePeriodSeconds": 60,
+                        "containers": [
+                            {
+                                "name": "${APPLICATION_NAME}-amq",
+                                "image": "jboss-amq-62",
```
Comment (tremes) [inline templates/eap-cd-amq-persistent-s2i.json]: Same as above

File: templates/eap-cd-amq-s2i.json [hunk 7]
```diff
@@ -0,0 +1,876 @@
...
+            "value": "eap-app",
+            "required": true
+        },
+        {
+            "displayName": "Custom http Route Hostname",
+            "description": "Custom hostname for http service route.  Leave blank for default hostname, e.g.: <application-name>-<project>.<default-domain-suffix>",
+            "name": "HOSTNAME_HTTP",
+            "value": "",
+            "required": false
+        },
```
Comment (tremes) [inline templates/eap-cd-amq-s2i.json]: This could be removed as well.

File: templates/eap-cd-amq-s2i.json [hunk 8]
```diff
@@ -0,0 +1,876 @@
...
+                        "type": "ImageChange",
+                        "imageChangeParams": {
+                            "automatic": true,
+                            "containerNames": [
+                                "${APPLICATION_NAME}-amq"
+                            ],
+                            "from": {
+                                "kind": "ImageStreamTag",
+                                "namespace": "${IMAGE_STREAM_NAMESPACE}",
+                                "name": "jboss-amq-62:1.7"
```
Comment (tremes) [inline templates/eap-cd-amq-s2i.json]: This should be updated to AMQ 6.3 and the same in all remaining AMQ templates.

File: templates/eap-cd-basic-s2i.json [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+            "description": "Git source URI for application",
+            "name": "SOURCE_REPOSITORY_URL",
+            "value": "https://github.com/jboss-developer/jboss-eap-quickstarts.git",
+            "required": true
+        },
+        {
+            "displayName": "Git Reference",
+            "description": "Git branch/tag reference",
+            "name": "SOURCE_REPOSITORY_REF",
+            "value": "7.1.0.GA",
```
Comment (tremes) [inline templates/eap-cd-basic-s2i.json]: This should currently point to "openshift" branch but it will likely change in the future.

**Comment by tremes** [comment]
`eap-cd-postgresql-*.json` templates somehow lost readinessProbe and livenessProbe definitions.

File: templates/eap-cd-basic-s2i.json (hunk 9 above)
Comment (luck3y) [reply templates/eap-cd-basic-s2i.json]: @tremes Yep, this one needs final testing with the openshift branch. I'll update when I've tested.

File: templates/eap-cd-amq-persistent-s2i.json (hunk 5 above)
Comment (luck3y) [reply templates/eap-cd-amq-persistent-s2i.json]: Will use jboss-amq-63:1.4

**Comment by luck3y** [comment]
@bstansberry @rcernich Can you guys have a final look at this? I'd like to merge it to the eap-cd-dev branch ASAP today.

Thanks!

File: image.yaml [hunk 10]
```diff
@@ -110,17 +114,19 @@ modules:
...
           - name: os-eap7-modules
-          - name: os-eap71-modules
+          - name: jboss.eap.cd.openshift.modules
           - name: os-eap7-ping
           - name: os-eap-activemq-rar
           - name: os-java-run
           - name: os-eap-launch
           - name: os-eap7-launch
+          - name: jboss.eap.cd.openshift.launch
+          - name: os-logging
```
Comment (bstansberry) [inline image.yaml]: This module is repeated below.

Comment (luck3y) [reply image.yaml]: Ah, yeah, os-logging.

Comment (luck3y) [reply image.yaml]: @bstansberry fixed to match the eap71 order.

File: templates/eap-cd-amq-persistent-s2i.json [hunk 11]
```diff
@@ -0,0 +1,924 @@
...
+    },
+    "message": "A new JBoss EAP CD and A-MQ persistent based application with SSL support has been created in your project. The username/password for accessing the A-MQ service is ${MQ_USERNAME}/${MQ_PASSWORD}. Please be sure to create the following secrets: \"${HTTPS_SECRET}\" containing the ${HTTPS_KEYSTORE} file used for serving secure content; \"${JGROUPS_ENCRYPT_SECRET}\" containing the ${JGROUPS_ENCRYPT_KEYSTORE} file used for securing JGroups communications.",
+    "parameters": [
+        {
+            "displayName": "Application Name",
+            "description": "The name for the application.",
+            "name": "APPLICATION_NAME",
+            "value": "eap-app",
+            "required": true
+        },
```
Comment (bstansberry) [inline templates/eap-cd-amq-persistent-s2i.json:30]: The 7.1 template also has: https://github.com/jboss-openshift/application-templates/blob/master/eap/eap71-amq-persistent-s2i.json#L31

Comment (bstansberry) [reply templates/eap-cd-amq-persistent-s2i.json:30]: It seems this was removed per a comment from Tomas Remes. So ok. :)

File: templates/eap-cd-amq-persistent-s2i.json [hunk 12]
```diff
@@ -0,0 +1,924 @@
...
+            "metadata": {
+                "name": "${APPLICATION_NAME}",
+                "labels": {
+                    "application": "${APPLICATION_NAME}"
+                },
+                "annotations": {
+                    "description": "Route for application's HTTP service."
+                }
+            },
+            "spec": {
```
Comment (bstansberry) [inline templates/eap-cd-amq-persistent-s2i.json:399]: The 7.1 template also has https://github.com/jboss-openshift/application-templates/blob/master/eap/eap71-amq-persistent-s2i.json#L407

Comment (bstansberry) [reply templates/eap-cd-amq-persistent-s2i.json:399]: Never mind if this was removed because the section I commented on above was removed

File: templates/eap-cd-basic-s2i.json (hunk 9 above)
Comment (bstansberry) [reply templates/eap-cd-basic-s2i.json]: +1

File: templates/eap-cd-image-stream.json [hunk 13]
```diff
@@ -0,0 +1,87 @@
+{
+    "kind": "List",
+    "apiVersion": "v1",
+    "metadata": {
+        "name": "jboss-cd-image-streams",
+        "annotations": {
+            "description": "ImageStream definitions for JBoss Enterprise Application Platform Continuous Delivery.",
```
Comment (bstansberry) [inline templates/eap-cd-image-stream.json]: I'm unclear about "Continuous Delivery" vs "continuous delivery" throughout this file.

Also, in eap71-image-stream.json it's "definition" not "definitions". I've no idea what's right; probably both are since 'items' is a single element array. :)

Comment (luck3y) [reply templates/eap-cd-image-stream.json]: As far as I'm aware these should all be lower case at this point, I'll update to be consistent.

File: templates/eap-cd-amq-persistent-s2i.json (hunk 11 above)
Comment (luck3y) [reply templates/eap-cd-amq-persistent-s2i.json:30]: Yep, HOSTNAME_HTTP was dropped from the route spec, so then removed as something the user can set. This is a general issue, not just EAP CD.

File: templates/eap-cd-amq-persistent-s2i.json (hunk 12 above)
Comment (luck3y) [reply templates/eap-cd-amq-persistent-s2i.json:399]: Same comment as HOSTNAME_HTTP above.

File: templates/eap-cd-postgresql-persistent-s2i.json [hunk 14]
```diff
@@ -0,0 +1,881 @@
+{
+    "kind": "Template",
+    "apiVersion": "v1",
+    "metadata": {
+        "annotations": {
+            "iconClass": "icon-eap",
+            "tags": "eap,javaee,java,jboss",
+            "version": "1.4.9",
```
Comment (bstansberry) [inline templates/eap-cd-postgresql-persistent-s2i.json]: I just noticed this when comparing this one to the 7.1 one but I suspect the same issue may apply to all the others. In 7.1 these "versions" are 1.4.10

https://github.com/jboss-openshift/application-templates/commit/09604c86eb78362a602944d23e7f3d8439a8f1f4

Comment (luck3y) [reply templates/eap-cd-postgresql-persistent-s2i.json]: @bstansberry Thanks for both noticing that and reminding me. 

@rcernich How are we planning to manage these versions? Do they need to match the version tagged with the other templates? (Those are currently 1.4.10.) Do these versions matter outside of this repo?

File: templates/eap-cd-sso-s2i.json [hunk 15]
```diff
@@ -0,0 +1,838 @@
...
+                            "name": "eap-cd-openshift:12"
+                        },
+                        "env": [
+                            {
+                                "name": "ARTIFACT_DIR",
+                                "value": "${ARTIFACT_DIR}"
+                            },
+                            {
+                                "name": "MAVEN_ARGS_APPEND",
+                                "value": "-Pjboss-community-repository"
```
Comment (bstansberry) [inline templates/eap-cd-sso-s2i.json]: Is this related to the QS parent pom thing? Is it needed in any of the others?

This is a diff vs the 7.1 template that I haven't seen in the others.

Comment (luck3y) [reply templates/eap-cd-sso-s2i.json]: @bstansberry I'm currently doing some final testing with this and the openshit QS. We'll either keep this change and drop the maven eap module, or vice versa. Also, yes, its related to the maven parent thing.

Comment (luck3y) [reply templates/eap-cd-sso-s2i.json]: Actually I'm going to drop this change from here until I'm done with my testing. If needed I'll open a final seperate PR with the QS / Maven pom related fixes.

**Review by bstansberry — APPROVED** [review]
Looks good.  Some more knowledgable eyes on eap-cd-image-stream.json would be good as I don't pretend to understand all those settings.

**Review by rcernich — APPROVED** [review]
LGTM
